tests/views/test_collection_exercise.py

```diff
@@ -535,39 +591,101 @@ def test_collection_exercise_view_seft_exercise_states(self, mock_request, mock_
+        mock_details.return_value = seft_collection_exercise_details
         response = self.client.get(f"/surveys/{short_name}/{period}", follow_redirects=True)
         self.assertEqual(response.status_code, 200)
         self.assertIn("Setting ready for live".encode(), response.data)
         self.assertNotIn("Set as ready for live".encode(), response.data)
 
         # Set the exercise state to ENDED
-        mock_details.return_value = seft_collection_exercise_details_ended
+        seft_collection_exercise_details["collection_exercise"]["state"] = "ENDED"
+        mock_details.return_value = seft_collection_exercise_details
         response = self.client.get(f"/surveys/{short_name}/{period}", follow_redirects=True)
         self.assertEqual(response.status_code, 200)
         self.assertIn("Ended".encode(), response.data)
 
+    @requests_mock.mock()
+    def test_collection_exercise_details_single_eq_collection_instrument(self, mock_request):
+        # Given I have a collection exercise with a single eq collection instrument (EQ)
+        self._mock_build_collection_exercise_details(mock_request)
+
+        # When I call build_collection_exercise_details
+        with self.app.app_context():
+            exercise_dict = build_collection_exercise_details("MBS", "000000", include_ci=True)
+
+        # Then the collection exercise has 1 key (EQ) and 1 value in the collection instrument
+        expected_output = {"EQ": self.eq_collection_instrument}
+        self.assertEquals(expected_output, exercise_dict["collection_instruments"])
+        self.assertEquals(len(exercise_dict["collection_instruments"]["EQ"]), 1)
+
+    @requests_mock.mock()
+    def test_collection_exercise_details_multiple_seft_collection_instruments(self, mock_request):
+        # Given I have multiple collection instruments of the same type (2 x SEFT)
+        eq_and_seft = self.seft_collection_instruments.copy()
+        self._mock_build_collection_exercise_details(mock_request, False)
+        mock_request.get(f"{url_get_collection_instrument}?{ci_search_string}", json=eq_and_seft, complete_qs=True)
+
+        # When I call build_collection_exercise_details
+        with self.app.app_context():
+            exercise_dict = build_collection_exercise_details("MBS", "000000", include_ci=True)
+
+        # Then the collection exercise has 1 key (SEFT) and 2 values in the collection instrument
+        expected_output = {"SEFT": eq_and_seft}
+
+        self.assertEquals(expected_output, exercise_dict["collection_instruments"])
+        self.assertEquals(len(exercise_dict["collection_instruments"]["SEFT"]), 2)
+        self.assertEquals(exercise_dict["collection_instruments"]["SEFT"][0]["id"], collection_instrument_id)
+        self.assertEquals(exercise_dict["collection_instruments"]["SEFT"][1]["id"], collection_instrument_id_2)
+
+    @requests_mock.mock()
+    def test_collection_exercise_details_eq_and_seft_collection_instruments(self, mock_request):
```

Comment:
You could argue that some of these tests aren't needed and we could just test the endpoint, but that isn't particularly reliable, for things like count we have nothing really to assert against bar a td class which isn't great, we should always aim to assert dict contents and not response.data, unless naturally test the display

Reply:
Sounds good to me.  I was going to say this should probably be in `controllers/test_collection_exercise_controllers.py` but then realised this function lives in the view so I'd say this test is in the right place